agent/agent.go

```diff
@@ -682,205 +711,272 @@ func (a *agent) fetchServiceBannerLoop(ctx context.Context, aAPI proto.DRPCAgent
+			a.setLifecycle(codersdk.WorkspaceAgentLifecycleStarting)
+
+			// Perform overrides early so that Git auth can work even if users
+			// connect to a workspace that is not yet ready. We don't run this
+			// concurrently with the startup script to avoid conflicts between
+			// them.
+			if manifest.GitAuthConfigs > 0 {
+				// If this fails, we should consider surfacing the error in the
+				// startup log and setting the lifecycle state to be "start_error"
+				// (after startup script completion), but for now we'll just log it.
+				err := gitauth.OverrideVSCodeConfigs(a.filesystem)
+				if err != nil {
+					a.logger.Warn(ctx, "failed to override vscode git auth configs", slog.Error(err))
+				}
+			}
+
+			err = a.scriptRunner.Init(manifest.Scripts)
+			if err != nil {
+				return xerrors.Errorf("init script runner: %w", err)
+			}
+			err = a.trackGoroutine(func() {
+				start := time.Now()
+				// here we use the graceful context because the script runner is not directly tied
+				// to the agent API.
+				err := a.scriptRunner.Execute(a.gracefulCtx, func(script codersdk.WorkspaceAgentScript) bool {
+					return script.RunOnStart
+				})
+				// Measure the time immediately after the script has finished
+				dur := time.Since(start).Seconds()
+				if err != nil {
+					a.logger.Warn(ctx, "startup script(s) failed", slog.Error(err))
+					if errors.Is(err, agentscripts.ErrTimeout) {
+						a.setLifecycle(codersdk.WorkspaceAgentLifecycleStartTimeout)
+					} else {
+						a.setLifecycle(codersdk.WorkspaceAgentLifecycleStartError)
+					}
+				} else {
+					a.setLifecycle(codersdk.WorkspaceAgentLifecycleReady)
+				}
+
+				label := "false"
+				if err == nil {
+					label = "true"
+				}
+				a.metrics.startupScriptSeconds.WithLabelValues(label).Set(dur)
+				a.scriptRunner.StartCron()
+			})
+			if err != nil {
+				return xerrors.Errorf("track conn goroutine: %w", err)
+			}
 		}
 		return nil
-	})
+	}
+}
 
-	eg.Go(func() error {
-		a.logger.Debug(egCtx, "running stats report loop")
-		err := a.statsReporter.reportLoop(egCtx, aAPI)
-		if err != nil {
-			return xerrors.Errorf("report stats loop: %w", err)
+// createOrUpdateNetwork waits for the manifest to be set using manifestOK, then creates or updates
+// the tailnet using the information in the manifest
+func (a *agent) createOrUpdateNetwork(manifestOK <-chan struct{}, networkOK chan<- struct{}) func(context.Context, drpc.Conn) error {
+	return func(ctx context.Context, _ drpc.Conn) error {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-manifestOK:
+		}
+		var err error
+		manifest := a.manifest.Load()
+		a.closeMutex.Lock()
+		network := a.network
+		a.closeMutex.Unlock()
+		if network == nil {
+			// use the graceful context here, because creating the tailnet is not itself tied to the
+			// agent API.
+			network, err = a.createTailnet(a.gracefulCtx, manifest.AgentID, manifest.DERPMap, manifest.DERPForceWebSockets, manifest.DisableDirectConnections)
```

Comment:
I feel like this context may be problematic during `Close`. If we close the agent before we've established a connection, then the connection can't establish and we have some connection dependent things in `Close`.

Why should we limit tailnet to anything except "hardCtx"?

This function is also pretty confusing since it takes `ctx` as an argument, but uses graceful here.

Reply:
Yeah, this whole thing needs a bigger refactor like you suggested above.  This routine doesn't directly use the `drpc.Conn` at all, so could arguably be outside the `apiConnRoutineManager` --- but it does rely on the manifest, and several other routines block on the network being available.  Fixing that is more than I wanted to bite off.

In terms of the context, I don't really think it matters whether we use the `gracefulCtx` or the `hardCtx`. As soon as we cancel the `gracefulCtx`, then we send disconnect to the coordinator and clients will start disconnecting their tunnels.

If we close before we've _ever_ established a tailnet, the I don't think anything in `Close()` can block on it. We'd never have any SSH sessions, and would never have connected to the Coordinator.